Unittest.py

```diff
@@ -68,6 +80,164 @@ def load_spoiler(json_file):
+                    if location in spoiler['locations']:
+                        item = spoiler['locations'][location]
+                        if isinstance(item, dict) and item['item'] == "Ice Trap":
+                            self.assertIn("model", item)
+                        else:
+                            self.assertNotIn("Ice Trap", item)
+
+    def test_should_not_throw_exception(self):
+        filenames = [
+            "plando-bottles-in-list",
+            "plando-bottle-item-group",
+            "plando-bottle-item-group-in-list",
+            "plando-adult-trade-in-list",
+            "plando-adult-trade-item-group",
+            "plando-adult-trade-item-group-in-list",
+            "plando-weird-egg-in-list",
+            "plando-shop-items",
+            "plando-list-case-sensitivity",
+            "plando-num-adult-trade-item-good",
+            "plando-num-weird-egg-item-good",
+            "plando-num-bottles-fountain-closed-good",
+            "plando-num-bottles-fountain-open-good"
+        ]
+        for filename in filenames:
+            with self.subTest(filename):
+                generate_with_plandomizer(filename)
+
+    def test_boss_item_list(self):
+        filenames = [
+            "plando-boss-list-child",
+            "plando-boss-list-adult",
+            "plando-boss-list"]
+        for filename in filenames:
+            with self.subTest(filename):
+                distribution_file, spoiler = generate_with_plandomizer(filename)
+                for location, item_list in distribution_file['locations'].items():
+                    self.assertIn(spoiler['locations'][location], item_list)
+
+    def test_pool_accuracy(self):
+        filenames = [
+            "empty",
+            "plando-list",
+            "plando-item-pool-matches-items-placed-after-starting-items-replaced"
+        ]
+        for filename in filenames:
+            with self.subTest(filename + " pool accuracy"):
+                distribution_file, spoiler = generate_with_plandomizer(filename)
+                self.check_pool_accuracy(spoiler, spoiler['item_pool'])
+        filename = "plando-list-exhaustion"
+        with self.subTest(filename + " pool accuracy"):
+            distribution_file, spoiler = generate_with_plandomizer(filename)
+            self.check_pool_accuracy(spoiler, distribution_file['item_pool'])
+
+    def check_pool_accuracy(self, spoiler, pool):
+        """Ensures pool values match spoiler item distribution
+
+        :param spoiler: Spoiler log output from generator
+        :param pool: Item pool to use for verification
+        :return:
+        """
+        for location, item in spoiler['locations'].items():
+            if isinstance(item, dict):
+                test_item = item['item']
+            else:
+                test_item = item
+            if test_item in pool:
+                pool[test_item] -= 1
+                self.assertGreaterEqual(pool[test_item], 0)
```

Comment:
I'm not a fan of this approach. How about constructing a pool from the spoiler locations that have items from the distribution pool? Then we can compare `pool[test_item]` to `actual[test_item]` and do the assert and error message on that.

Reply:
I'm not sure if your way will handle the various behaviours I am testing with this.

Pretty much all cases are attempting to test that my item pool updating in Plandomizer.py doesn't cause the pool to be inaccurate.

Testing the item pool of an Empty list is likely worthless, I'll probably remove that.

The plando-list test is the most likely to work properly using your proposed method without modification or assumptions, and is the next least likely to be worthwhile to run through this function in the first place.

For the one with starting_items, all starting_items are removed from the item pool and replaced with `#Junk`. These do not appear as locations in the distribution file, they just are junk items added to the pool to be randomly placed. What this code is testing, at least for that test in particular, is that the junk entries in the item_pool reflect the actual number of that item in the spoiler log. My code this test is for would replace the junk value with 1 and count up from there, or just repeatedly set the value to 1 in previous iterations. It also tests that deleting a key from the item_pool works when the value reaches 0 as happens with all of the starting_items.

And plando-list-exhaustion is testing that the items in the distribution file are the only ones of those kind placed in the world. Once 3 bows, 2 magic meters, and 1 light arrows have been placed the item pool no others of those items should be placed in the world. If we can assume the starting_items test properly tests that the item pool updating code is working properly I suppose this could be a separate test using your method, or removed.

So I could likely at least reduce this down to being used in one test and not have it as a separate function.

Reply:
You're grabbing all the placed items (locations in the spoiler file), and removing them one-by-one from the pool that you've passed to this function, and asserting the counts don't go below 0. Instead, I've suggested just counting over all the placed items and then comparing them to the initial counts. This would get us more descriptive failure messages, like `pool had 5 #Junk items but 10 #Junk placed`, and should be functionally equivalent, except for

> It also tests that deleting a key from the item_pool works when the value reaches 0 as happens with all of the starting_items.

which means that all you're actually testing is that the counts in the item pool are positive. (The test that the item_pool deletes the key when it reaches 0 should be a separate test. Here's it's implicit: the tests will only fail if you've placed more items than were in the pool, which is the main point of the test.)

Reply:
Okay, I think I understand now. I'll work on changing to that.

Reply:
So is something like this more in line with what you want?
```Python
actual_pool = {}
for location, item in spoiler['locations'].items():
    if isinstance(item, dict):
        test_item = item['item']
    else:
        test_item = item
    try:
        actual_pool[test_item] += 1
    except KeyError:
        actual_pool[test_item] = 1
for item in pool:
    self.assertEqual(actual_pool[item], pool[item])
```

Reply:
Yeah, that's easier to understand what you want. And you can actually condense the last for loop into `self.assertEqual(actual_pool, pool)`.

Reply:
Will that actually work? Actual pool will have things like Maps/Compasses/Small Keys outside of keysanity, and other things in it.